synapse/crypto/keyring.py

```diff
@@ -903,3 +914,31 @@ async def get_server_verify_keys_v2_direct(
             response_json=response,
             time_added_ms=time_now_ms,
         )
+
+
+class InternalWorkerRequestKeyFetcher(KeyFetcher):
+    """Ask a federation_sender worker to request keys for some homeserver X."""
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__(hs)
+        self._federation_shard_config = hs.config.worker.federation_shard_config
+        self._client = ReplicationFetchKeysEndpoint.make_client(hs)
+
+    async def _fetch_keys(
+        self, keys_to_fetch: List[_FetchKeyRequest]
+    ) -> Dict[str, Dict[str, FetchKeyResult]]:
+        # For simplicity's sake, pick a random federation sender
```

Comment:
I think we would want to choose the same federation sender that's talking to that instance? 🤷

Reply:
That's what I thought, but `federation_shard_config` is a `ShardedWorkerHandlingConfig` rather than a `class RoutableShardedWorkerHandlingConfig`; only the latter defines `get_instance` publicly.

Personally I find this all incomprehensible.